File: platform/providers/mock.js
/**
 * The simulated target LLM.
 *
 * It is deliberately vulnerable in exactly the way real tool-using LLMs are:
 * it flattens every piece of text it sees -- the user's message AND the output
 * of the tools it calls -- into one undifferentiated instruction stream.
 * Guardrails are only ever applied to the *user* channel, which is precisely
 * why indirect injection works: content arriving via a tool result is never
 * re-screened, but is still obeyed.
 */

const MAX_STEPS = 6;

export function createMockProvider(scenario) {
  return {
    name: 'mock',
    label: 'Simulated LLM (offline, deterministic)',

    async run({ userMessage, state, history }) {
      const trace = [];
      const said = [];
      const calledTools = [];

      // Each "source" is a chunk of text the model has in its context window.
      // channel: 'user'   -> arrived over the trusted-looking chat channel
      // channel: 'tool'   -> arrived inside a tool result (attacker-controllable)
      const sources = [{ text: userMessage, channel: 'user', origin: 'user message' }];
      trace.push({ kind: 'user', from: 'user', to: 'LLM', text: userMessage });

      let step = 0;
      let refused = false;
      const fired = new Set(); // a tool can't be re-triggered by its own output

      while (sources.length && step < MAX_STEPS) {
        const src = sources.shift();
        step++;

        // --- Guardrails: applied to the user channel ONLY. This is the bug. ---
        if (src.channel === 'user') {
          const verdict = screen(scenario.guardrails || [], src.text);
          if (verdict?.blocked) {
            trace.push({
              kind: 'guardrail',
              status: 'blocked',
              guardrail: verdict.id,
              text: `Guardrail "${verdict.id}" refused this request.`,
            });
            said.push(verdict.refusal);
            refused = true;
            break;
          }
          if (verdict?.bypassed) {
            trace.push({
              kind: 'guardrail',
              status: 'bypassed',
              guardrail: verdict.id,
              technique: verdict.technique,
              text: `Guardrail "${verdict.id}" was bypassed via ${verdict.technique}.`,
            });
          }
        } else {
          trace.push({
            kind: 'trust',
            text: `Content from ${src.origin} is being read as instructions. It was never screened by the guardrails — only the user channel is.`,
          });
        }

        // --- Intent matching: which tools does this text ask for? ---
        const matches = matchIntents(scenario.tools, src.text).filter((m) => !fired.has(m.tool.name));
        for (const { tool, args } of matches) {
          fired.add(tool.name);
          trace.push({
            kind: 'call',
            from: 'LLM',
            to: 'API',
            text: `${tool.name}(${fmtArgs(args)})`,
            injected: src.channel === 'tool',
          });

          let result;
          try {
            result = await tool.run(args, state) || {};
          } catch (err) {
            result = { say: `The ${tool.name} call failed: ${err.message}` };
          }

          calledTools.push({ name: tool.name, args, result });
          trace.push({
            kind: 'result',
            from: 'API',
            to: 'LLM',
            text: result.raw ?? result.say ?? '(no output)',
            tainted: Boolean(result.tainted),
          });

          if (result.say) said.push(result.say);

          // The returned text goes straight back into the context window and is
          // scanned for further instructions -- the indirect injection path.
          if (result.raw) {
            sources.push({
              text: result.raw,
              channel: 'tool',
              origin: `the ${tool.name} tool result`,
            });
          }
        }
      }

      if (!refused && !said.length) {
        said.push(answerFromKnowledge(scenario, userMessage, history));
      }

      const reply = said.filter(Boolean).join('\n\n');
      trace.push({ kind: 'reply', from: 'LLM', to: 'user', text: reply });
      return { reply, trace, calledTools };
    },
  };
}

/* ------------------------------------------------------------------ */

function screen(guardrails, text) {
  for (const g of guardrails) {
    const blocks = toArray(g.blocks);
    if (!blocks.some((re) => re.test(text))) continue;

    for (const b of g.bypasses || []) {
      if (b.re.test(text)) return { id: g.id, bypassed: true, technique: b.name };
    }
    return { id: g.id, blocked: true, refusal: g.refusal };
  }
  return null;
}


function matchIntents(tools, text) {
  const out = [];
  for (const tool of tools) {
    for (const intent of tool.intents || []) {
      const m = intent.re.exec(text);
      if (!m) continue;
      const args = intent.args ? intent.args(m, text) : {};
      out.push({ tool, args, index: m.index });
      break; // one call per tool per source
    }
  }
  return out.sort((a, b) => a.index - b.index);
}

function answerFromKnowledge(scenario, userMessage, history) {
  for (const k of scenario.knowledge || []) {
    if (k.re.test(userMessage)) return typeof k.say === 'function' ? k.say(history) : k.say;
  }
  return scenario.fallbackReply || "I'm not sure I can help with that. Try asking about a product or an order.";
}

function fmtArgs(args) {
  return Object.entries(args || {})
    .map(([k, v]) => `${k}=${JSON.stringify(v)}`)
    .join(', ');
}

function toArray(v) {
  return Array.isArray(v) ? v : [v];
}

File: backend/providers/mock.py
"""The simulated target LLM — deliberately vulnerable in the same way real
tool-using models are: it flattens the user message AND tool output into one
instruction stream, and only ever screens the *user* channel. Content arriving
via a tool result is never re-screened but is still obeyed → indirect injection.
"""
import json

MAX_STEPS = 6


class MockProvider:
    name = "mock"
    label = "Simulated LLM (offline, deterministic)"

    def __init__(self, scenario):
        self.scenario = scenario

    def run(self, user_message, state, history):
        scenario = self.scenario
        trace = []
        said = []
        called_tools = []

        # Each source is a chunk of text in the context window.
        #   channel 'user' -> arrived over the (screened) chat channel
        #   channel 'tool' -> arrived inside a tool result (attacker-controllable)
        sources = [{"text": user_message, "channel": "user", "origin": "user message"}]
        trace.append({"kind": "user", "from": "user", "to": "LLM", "text": user_message})

        step = 0
        refused = False
        fired = set()  # a tool can't be re-triggered by its own output

        while sources and step < MAX_STEPS:
            src = sources.pop(0)
            step += 1

            if src["channel"] == "user":
                verdict = _screen(scenario.get("guardrails", []), src["text"])
                if verdict and verdict.get("blocked"):
                    trace.append(
                        {
                            "kind": "guardrail",
                            "status": "blocked",
                            "guardrail": verdict["id"],
                            "text": f"Guardrail \"{verdict['id']}\" refused this request.",
                        }
                    )
                    said.append(verdict["refusal"])
                    refused = True
                    break
                if verdict and verdict.get("bypassed"):
                    trace.append(
                        {
                            "kind": "guardrail",
                            "status": "bypassed",
                            "guardrail": verdict["id"],
                            "technique": verdict["technique"],
                            "text": f"Guardrail \"{verdict['id']}\" was bypassed via {verdict['technique']}.",
                        }
                    )
            else:
                trace.append(
                    {
                        "kind": "trust",
                        "text": (
                            f"Content from {src['origin']} is being read as instructions. "
                            "It was never screened by the guardrails — only the user channel is."
                        ),
                    }
                )

            matches = [m for m in _match_intents(scenario["tools"], src["text"]) if m["tool"]["name"] not in fired]
            for match in matches:
                tool = match["tool"]
                args = match["args"]
                fired.add(tool["name"])
                trace.append(
                    {
                        "kind": "call",
                        "from": "LLM",
                        "to": "API",
                        "text": f"{tool['name']}({_fmt_args(args)})",
                        "injected": src["channel"] == "tool",
                    }
                )
                try:
                    result = tool["run"](args, state) or {}
                except Exception as err:  # noqa: BLE001
                    result = {"say": f"The {tool['name']} call failed: {err}"}

                called_tools.append({"name": tool["name"], "args": args, "result": result})
                trace.append(
                    {
                        "kind": "result",
                        "from": "API",
                        "to": "LLM",
                        "text": result.get("raw", result.get("say", "(no output)")),
                        "tainted": bool(result.get("tainted")),
                    }
                )
                if result.get("say"):
                    said.append(result["say"])
                if result.get("raw"):
                    sources.append({"text": result["raw"], "channel": "tool", "origin": f"the {tool['name']} tool result"})

        if not refused and not said:
            said.append(_answer_from_knowledge(scenario, user_message, history))

        reply = "\n\n".join([s for s in said if s])
        trace.append({"kind": "reply", "from": "LLM", "to": "user", "text": reply})
        return {"reply": reply, "trace": trace, "calledTools": called_tools}


def _screen(guardrails, text):
    for g in guardrails:
        blocks = g["blocks"] if isinstance(g["blocks"], list) else [g["blocks"]]
        if not any(re.search(text) for re in blocks):
            continue
        for b in g.get("bypasses", []):
            if b["re"].search(text):
                return {"id": g["id"], "bypassed": True, "technique": b["name"]}
        return {"id": g["id"], "blocked": True, "refusal": g["refusal"]}
    return None


def _match_intents(tools, text):
    out = []
    for tool in tools:
        for intent in tool.get("intents", []):
            m = intent["re"].search(text)
            if not m:
                continue
            args = intent["args"](m, text) if intent.get("args") else {}
            out.append({"tool": tool, "args": args, "index": m.start()})
            break  # one call per tool per source
    out.sort(key=lambda x: x["index"])
    return out


def _answer_from_knowledge(scenario, user_message, history):
    for k in scenario.get("knowledge", []):
        if k["re"].search(user_message):
            say = k["say"]
            return say(history) if callable(say) else say
    return scenario.get("fallbackReply", "I'm not sure I can help with that.")


def _fmt_args(args):
    return ", ".join(f"{k}={json.dumps(v)}" for k, v in (args or {}).items())
